test/AutoDiff/leakchecking.swift

```diff
@@ -23,19 +28,89 @@ struct ExampleLeakModel : Differentiable {
   }
 }
 
+struct FloatPair : Differentiable & AdditiveArithmetic {
+  var first, second: Tracked<Float>
+  init(_ first: Tracked<Float>, _ second: Tracked<Float>) {
+    self.first = first
+    self.second = second
+  }
+}
+
+struct Pair<T : Differentiable, U : Differentiable> : Differentiable
+  where T == T.AllDifferentiableVariables, T == T.TangentVector,
+        U == U.AllDifferentiableVariables, U == U.TangentVector
+{
+  var first: Tracked<T>
+  var second: Tracked<U>
+  init(_ first: Tracked<T>, _ second: Tracked<U>) {
+    self.first = first
+    self.second = second
+  }
+}
+
 LeakCheckingTests.test("BasicVarLeakChecking") {
-  do {
+  testWithLeakChecking(expectedLeakCount: 0) {
+    var model = ExampleLeakModel()
+    let x: Tracked<Float> = 1.0
+    _ = model.gradient(at: x) { m, x in m.applied(to: x) }
+  }
+
+  // FIXME: Fix closure capture related memory leak.
+  testWithLeakChecking(expectedLeakCount: 1) {
```

Comment:
Oops - this test actually has a leak count of 0.
But I recall encountering a test like this that fails with a leak count of 1.

cc @rxwei @pschuh: do you recall any reproducers for the known closure capture related memory leak?

Reply:
Looks like it will leak.  Try wrapping things in a loop?
```swift
for _ in 0..<10 {
  let x: Tracked<Float> = 1.0
  _ = model.gradient { m in m.applied(to: x) }
}
```

Reply:
Added more tests in https://github.com/apple/swift/pull/25249/commits/d4691b878951a0f26bcfd5033ceb30937d33128e.

Loop actually doesn't leak, but I found a test case that does leak.
```swift
  // TODO: Fix memory leak.
  testWithLeakChecking(expectedLeakCount: 1) {
    var model = ExampleLeakModel()
    let x: Tracked<Float> = 1.0
    _ = model.gradient { m in
      var model = m
      // Next line causes leak.
      model.bias = x
      return model.applied(to: x)
    }
  }
```

[Full dump here](https://gist.github.com/dan-zheng/27c9a3a612bf144df00b0f6bf110164a), in case you have time to look into it along with control flow related memory leaks.

Reply:
Interesting. This doesn't seem related to closure checking then. What if you wrote:

```swift
    _ = model.gradient { m in
      let x: Tracked<Float> = 1.0
      var model = m
      // Next line causes leak.
      model.bias = x
      return model.applied(to: x)
    }
```

Reply:
> Interesting. This doesn't seem related to closure checking then. What if you wrote: ...

The partial snippet you shared also produces a leak (it's pretty similar to the leak that's checked in). Here's the full snippet:
```swift
  // TODO: Fix memory leak.
  testWithLeakChecking(expectedLeakCount: 1) {
    var model = ExampleLeakModel()
    _ = model.gradient { m in
      let x: Tracked<Float> = 1.0
      var model = m
      // Next line causes leak.
      model.bias = x
      return model.applied(to: x)
    }
  }
```

Reply:
Yeah, so it should be unrelated to closure captures. It's a mutation differentiation bug. Could you file a bug for this specific issue with a reproducer?

Reply:
Filed [TF-550](https://bugs.swift.org/browse/TF-550).

Reply:
I think dan's example here of adding tests with expectedLeakCount is a better way to keep track of the leaks and their reproducers. Checking the leaks in would keep track of when they are fixed better. The bug could also contain the reproducer.